tests/test_network.py

```diff
@@ -0,0 +1,553 @@
+        assert len(requests) == 1
+
+    @sync
+    @chrome_only
+    async def test_fires_for_iframes(self, server, isolated_page):
+        """The page event should be fired for iframes."""
+        page = isolated_page
+        requests = []
+        page.on('request', lambda request: listener(requests, request))
+        await page.goto(server.empty_page)
+        await utils.attachFrame(page, 'frame1', server.empty_page)
+        assert len(requests) == 2
+
+    @sync
+    async def test_fires_for_fetches(self, server, isolated_page):
+        """The page event should be fired for fetches."""
+        page = isolated_page
+        requests = []
+        page.on('request', lambda request: listener(requests, request))
+        await page.goto(server.empty_page)
+        await page.evaluate("fetch('/empty.html')")
+        assert len(requests) == 2
+
+
+class TestRequestFrame:
+
+    @sync
+    async def test_works_for_main_frame_nav_request(self, server, isolated_page):
+        """Request frame should work for main frame navigation request."""
+        page = isolated_page
+        requests = []
+        page.on('request', lambda request: listener(requests, request))
+        await page.goto(server.empty_page)
+        assert len(requests) == 1
+        assert requests[0].frame == page.mainFrame
+
+    @sync
+    @chrome_only
+    async def test_works_for_subframe_nav_request(self, server, isolated_page):
+        """Request frame should work for subframe navigation request."""
+        page = isolated_page
+        requests = []
+        page.on('request', lambda request: listener(requests, request))
+        await page.goto(server.empty_page)
+        await utils.attachFrame(page, 'frame1', server.empty_page)
+        assert len(requests) == 2
+        assert requests[0].frame == page.frames[0]
+
+    @sync
+    @chrome_only
+    async def test_works_for_fetch_requests(self, server, isolated_page):
+        """Request frame should work for fetch requests."""
+        page = isolated_page
+        await page.goto(server.empty_page)
+
+        requests = []
+        page.on('request', lambda request: listener(requests, request))
+        await page.evaluate("fetch('/digits/1.png')")
+        assert len(requests) == 1
+        assert requests[0].frame == page.mainFrame
+
+
+class TestRequestHeader:
+
+    @sync
+    @chrome_only
+    async def test_header_contains_user_agent(self, server, isolated_page):
+        """Verify request contains a info about user agent in header."""
+        page = isolated_page
+        response = await page.goto(server.empty_page)
+        assert 'Chrome' in response.request.headers['user-agent']
+
+
+class TestResponseHeader:
+
+    @sync
+    @chrome_only
+    async def test_header_contains_info(self, server, isolated_page):
+        """Verify response header contains expected info."""
+        page = isolated_page
+        # server.setRoute('/empty.html', (req, res) = > {
```

Comment:
@Mattwmaster58 Desperately need help here: not sure how to handle this via the test server.

Reply:
Sec

Reply:
See the matrix room